sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala

```diff
@@ -88,6 +88,276 @@ case class Average(child: Expression) extends AlgebraicAggregate {
+  protected val currentM2 = AttributeReference("currentM2", resultType)()
+  protected val currentM1 = AttributeReference("currentM1", resultType)()
+  protected val currentM0 = AttributeReference("currentM0", resultType)()
+
+  override val bufferAttributes = List(currentM4, currentM3, currentM2, currentM1, currentM0)
+
+  override val initialValues = Seq(
+    /* currentM4 = */ Cast(Literal(0), resultType),
+    /* currentM3 = */ Cast(Literal(0), resultType),
+    /* currentM2 = */ Cast(Literal(0), resultType),
+    /* currentM1 = */ Cast(Literal(0), resultType),
+    /* currentM0 = */ Cast(Literal(0), resultType)
+  )
+
+  override val updateExpressions = {
+    lazy val updateM0: Expression = {
+      Add(currentM0, Cast(Literal(1), resultType))
+    }
+    lazy val delta = Subtract(Cast(child, resultType), currentM1)
+    lazy val deltaN = Divide(delta, updateM0)
+
+    lazy val updateM1: Expression = if (maxMoment >= 1) {
+      Add(currentM1, Divide(delta, updateM0))
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM2: Expression = if (maxMoment >= 2) {
+      Add(currentM2, Multiply(deltaN * delta, Subtract(updateM0, Cast(Literal(1), resultType))))
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM3: Expression = if (maxMoment >= 3) {
+      currentM3 + deltaN * deltaN * delta * (updateM0 - Cast(Literal(1), resultType)) *
+        (updateM0 - Cast(Literal(2), resultType)) -
+        deltaN * currentM2 * Cast(Literal(3), resultType)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM4: Expression = if (maxMoment >= 4) {
+      currentM4 + deltaN * deltaN * deltaN * delta * (updateM0 - Cast(Literal(1), resultType)) *
+        (updateM0 * updateM0 - updateM0 * Cast(Literal(3), resultType) +
+          Cast(Literal(3), resultType)) + deltaN * deltaN * currentM2 *
+        Cast(Literal(6), resultType) - deltaN * currentM3 * Cast(Literal(4), resultType)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    Seq(
+      /* currentM4 = */ If(IsNull(child), currentM4, updateM4),
+      /* currentM3 = */ If(IsNull(child), currentM3, updateM3),
+      /* currentM2 = */ If(IsNull(child), currentM2, updateM2),
+      /* currentM1 = */ If(IsNull(child), currentM1, updateM1),
+      /* currentM0 = */ If(IsNull(child), currentM0, updateM0)
+    )
+  }
+
+  override val mergeExpressions = {
+
+
+    lazy val updateM0 = currentM0.left + currentM0.right
+    lazy val delta = currentM1.right - currentM1.left
+    lazy val deltaN = delta / updateM0
+
+    lazy val updateM1 = if (maxMoment >= 1) {
+      currentM1.left + delta * (currentM0.right / updateM0)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM2 = if (maxMoment >= 2) {
+      currentM2.left + currentM2.right + delta * deltaN * currentM0.left * currentM0.right
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM3 = if (maxMoment >= 3) {
```

Comment:
maxMoment variable is implemented with intent to avoid extra computation. For instance, if variance function calls `CentralMomentAgg` then the 3rd and 4th moments do not need to be computed. Would appreciate feedback/criticism of this implementation.